mobile/src/main/java/ch/epfl/sdp/mobile/state/game/delegating/DelegatingPuzzlePromotionMovableChessBoardState.kt

```diff
@@ -0,0 +1,58 @@
+package ch.epfl.sdp.mobile.state.game.delegating
+
+import ch.epfl.sdp.mobile.application.authentication.AuthenticatedUser
+import ch.epfl.sdp.mobile.application.chess.engine.Color
+import ch.epfl.sdp.mobile.application.chess.engine.NextStep
+import ch.epfl.sdp.mobile.application.chess.engine.rules.Action
+import ch.epfl.sdp.mobile.state.game.AbstractMovableChessBoardState
+import ch.epfl.sdp.mobile.state.game.core.MutableGameDelegate
+import ch.epfl.sdp.mobile.state.game.delegating.DelegatingChessBoardState.Companion.toEnginePosition
+import ch.epfl.sdp.mobile.state.game.delegating.DelegatingChessBoardState.Companion.toPosition
+import ch.epfl.sdp.mobile.state.game.delegating.DelegatingChessBoardState.Companion.toRank
+import ch.epfl.sdp.mobile.ui.game.ChessBoardState
+
+/**
+ * An implementation of [AbstractMovableChessBoardState] which uses the [DelegatingPromotionState]
+ * to trigger the promotion dialog, and the [DelegatingPuzzleInfoState] to check whose turn it is to
+ * play.
+ *
+ * @param user the current [AuthenticatedUser].
+ * @param delegate the underlying [MutableGameDelegate].
+ * @param puzzleState the [DelegatingPuzzleInfoState], providing access to the puzzle's state info.
+ * @param promotion the [DelegatingPromotionState], providing access to the promotion info.
+ */
+class DelegatingPuzzlePromotionMovableChessBoardState(
+    private val user: AuthenticatedUser,
+    private val delegate: MutableGameDelegate,
+    private val puzzleState: DelegatingPuzzleInfoState,
+    private val promotion: DelegatingPromotionState,
+) : AbstractMovableChessBoardState(delegate) {
+
+  override fun move(from: ChessBoardState.Position, to: ChessBoardState.Position) {
+    val available =
+        delegate
+            .game
+            .actions(from.toEnginePosition())
+            .filter { (it.from + it.delta)?.toPosition() == to }
+            .toList()
+    val step = delegate.game.nextStep as? NextStep.MovePiece ?: return
+    val userCurrentlyPlaying =
+        step.turn ==
+            when (puzzleState.puzzleInfo.playerColor) {
+              ChessBoardState.Color.White -> Color.White
+              ChessBoardState.Color.Black -> Color.Black
+            }
```

Comment:
There is probably an extension method `toEngineColor()` method that already does this.

Reply:
Fixed in https://github.com/epfl-SDP/android/pull/309/commits/92d56e85430c9cb49c4d55670acf626ea3b527d1